feat: add k8s:unique tag

yongruilin opened the pull request:
`+k8s:unique` tag to support uniqueness checks on atomic lists.

Ref: https://github.com/jpbetz/validation-gen/issues/131

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/each.go:
```diff
@@ -307,24 +370,34 @@ func (lv listValidator) GetValidations(context Context) (Validations, error) {
 		}
+		comment := ""
+		if lm.declaredAsSet {
+			comment = "listType=set requires unique values"
```
Given my comment above, you could just say "lists with set semantics require unique values"

thockin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/each.go:
```diff
@@ -241,6 +240,70 @@ func (lmktv listMapKeyTagValidator) Docs() TagDoc {
+	doc := TagDoc{
+		Tag:         utv.TagName(),
+		Scopes:      utv.ValidScopes().UnsortedList(),
+		Description: "Declares a list field's elements are unique.",
```
nitL: "Declares that a ..."

We should improve the docs here and listType to be clear on the difference and overlap

yongruilin replied:
I extend the description here.

thockin reviewed (commented):
This was just about as simple as I hoped for!

yongruilin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/each.go:
```diff
@@ -76,14 +78,16 @@ type listMetadata struct {
 	declaredAsAtomic bool
 	declaredAsSet    bool
 	declaredAsMap    bool
+	uniqueAsSet      bool
```
This looks better. I didn't come up with a suitable naming. I've updated the code.

yongruilin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/each.go:
```diff
@@ -70,20 +72,28 @@ func init() {
-	keyFields        []string // iff declaredAsMap
-	keyNames         []string // iff declaredAsMap
+	semantic  listSemantic
+	unique    listSemantic // if set, overrides semantic for uniqueness
```
I thought if lists without any listType are atomic?

thockin replied:
Do we want to perpetuate that in DV?  I think it would be great to lint that every list has a listType, so people actually think about it.  Same as optional/required.

yongruilin replied:
We can do that, so when we migrate a package, all the lists would be examined to have a listType (k8s:listType). 
And k8s:unique would always be coupled with k8s:listType=atomic.

thockin replied:
That con is a pretty big con, isn't it?  Any other tools which understand what listType means might also need to update (eventually).  @jpbetz -- I don't want to bite off too much here.

yongruilin replied:
The ultimate goal is to have a uniqueness check on atomic list. I'm onboard to add `k8s:unique` on `k8s:listType=atomic`, so that we don't disrupt the SSA too much. 
However, we still need to update the openapi to take k8s:unique into consideration so that we can pass this related check
```go
	if len(schema.XListMapKeys) > 0 {
		if schema.XListType == nil {
			allErrs.SchemaErrors = append(allErrs.SchemaErrors, field.Required(fldPath.Child("x-kubernetes-list-type"), "must be map if x-kubernetes-list-map-keys is non-empty"))
		} else if *schema.XListType != "map" {
			allErrs.SchemaErrors = append(allErrs.SchemaErrors, field.Invalid(fldPath.Child("x-kubernetes-list-type"), *schema.XListType, "must be map if x-kubernetes-list-map-keys is non-empty"))
		}
	}
```

jpbetz replied:
That feels reasonable to me.

yongruilin replied:
Thanks Joe. I prefer to stick with the current implementation.
- `listType`: atomic, set, map
- `unique`: set, map (only with listType=atomic) 

It maintains SSA compatibility (big win).  It's also explicit and clear about what each tag does.

thockin replied:
We have two choices to make -- what are the tags (which I think listType + unique works) and what are we storing in listMetadata here to make this PR as readable as possible.  That's the part I was less sure about.

yongruilin replied:
I reconsider it. It makes more sense to have the listmetadata with `ownership` and `semantic` fields.
We don't use "ownership" for generating validation, but it is good to keep it to prevent logic from leaking and potentially can use to determine per-item ratcheting.
```go
type listOwnership string

const (
    ownershipSingle listOwnership = "single" // from listType=atomic
    ownershipShared listOwnership = "shared" // from listType=set/map
)

type listSemantic string

const (
    semanticAtomic listSemantic = "atomic" // No uniqueness check
    semanticSet    listSemantic = "set"    // uniqueness check
    semanticMap    listSemantic = "map"    // uniqueness check based on key(s)
)

type listMetadata struct {
    ownership      listOwnership // For now we don't use it for generation.
    semantic       listSemantic
    keyFields      []string
    keyNames       []string
}
```

The tags relation to fields would be like this
| `k8s:listType` Tag | `k8s:unique` Tag | Resulting `ownership` | Resulting `semantic` |
| :----------------- | :--------------- | :-------------------- | :------------------- |
| `"atomic"`         | `""`    | `single`              | `atomic`             |
| `"set"`            | `""`    | `shared`              | `set`                |
| `"map"`            | `""`    | `shared`              | `map`                |
| `"atomic"`         | `"set"`          | `single`              | `set`                |
| `"atomic"`         | `"map"`          | `single`              | `map`                |

jpbetz replied:
Where is the latest discussion happening?

yongruilin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/item.go:
```diff
@@ -94,7 +94,7 @@ func (itv *itemTagValidator) GetValidations(context Context, tag codetags.Tag) (
 	// Fields inherit list metadata from typedefs, but not vice-versa.
 	// If we find no listMetadata then something is wrong.
-	if !ok || !listMeta.declaredAsMap || len(listMeta.keyFields) == 0 {
+	if !ok || listMeta.semantic != listMap || len(listMeta.keyFields) == 0 {
```
Added a test under tags/item to test it.

yongruilin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/each.go:
```diff
@@ -309,54 +381,58 @@ func (lv listValidator) GetValidations(context Context) (Validations, error) {
 		}
+		comment := ""
+		if lm.semantic == listSet {
+			comment = "lists with set semantics require unique values"
```
Updated.

yongruilin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/each.go:
```diff
@@ -309,54 +381,58 @@ func (lv listValidator) GetValidations(context Context) (Validations, error) {
+		matchArg := lm.makeListMapMatchFunc(nt.Elem)
+		comment := ""
+		if lm.semantic == listMap {
+			// comment = "listType=map requires unique keys"
```
Updated.

thockin commented:
I already find it hard to remember that "zeroOrOneOf" is in a different file from union.

How about we put these together and rtename to list.go ?

yongruilin commented:
I've moved list-specific code to list.go and implemented my proposed solution https://github.com/jpbetz/validation-gen/pull/166#discussion_r2292726037

yongruilin commented:
@thockin @jpbetz I think it is now worth another pass?

yongruilin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/tags/item/single_key/doc.go:
```diff
@@ -50,6 +50,13 @@ type Struct struct {
 	// +k8s:item(id: "typedef-target")=+k8s:validateFalse="item TypedefItems[id=typedef-target] 1"
 	// +k8s:item(id: "typedef-target")=+k8s:validateFalse="item TypedefItems[id=typedef-target] 2"
 	TypedefItems TypedefItemList `json:"typedefItems"`
```
Is the tests under `output_tests/tags/item/typedef/...` the one you're looking for?

jpbetz replied:
Great. I think that's a good test to have.

I'm probably doing a bad job explaining why I brought this up.

I noticed that we have defensive code that prevents duplicates of some tags from being set on a field.  And so I started wondering:  "Couldn't I just bypass this by putting one of the tags on the field and another of the tags on the typedef?"

jpbetz replied:
Oh yes, I don't mean "no duplicates" in general, but for the tags in this PR that prevent duplicate use when on a field (like type=set) I noticed that I could sidestep that restriction using typedefs.

thockin replied:
You should not be able to do duplicates that are in conflict, should we try to prevent identical dups?

Can you show an example?

yongruilin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/each.go:
```diff
@@ -243,6 +251,70 @@ func (lmktv listMapKeyTagValidator) Docs() TagDoc {
 	return doc
 }
 
+type uniqueTagValidator struct {
```
Moved it to list.go

jpbetz commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/list.go:
```diff
@@ -0,0 +1,426 @@
+		return Validations{}, fmt.Errorf("can only be used on list types (%s)", t.Kind)
+	}
+
+	lm := lttv.byPath[context.Path.String()]
```
What do the strings keys for `lttv.byPath` look like?   It looked to me like this is a global map that accumulates entries as validation-gen runs.  Can we have Some time like Pod add a "spec.x" field then later have another type like Deployment which also has "spec.x" (this is a made-up example) collide?

jpbetz replied:
OK, great.  Thanks.

thockin reviewed (commented):
When I suggested rename, I think I meant just to rename each.go -> list.go, not break it up.  Everything in these 3 files is highly coupled anyway.  This is fine, too. :)

yongruilin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/each.go:
```diff
@@ -481,26 +168,25 @@ func (evtv eachValTagValidator) getListValidations(fldPath *field.Path, t *types
-			matchArg = Identifier(validateDirectEqual)
-		} else {
-			matchArg = Identifier(validateSemanticDeepEqual)
+	if listMetadata != nil {
```
Yes. Since now it has key(s) we can locate the old-value easily, though it is "atomic".

yongruilin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/list.go:
```diff
@@ -0,0 +1,426 @@
+	case "set":
+		lm.ownership = ownershipShared
+		// If uniqueTagValidator has run, lm.semantic will be non-empty and non-atomic.
+		if lm.semantic != "" && lm.semantic != semanticAtomic {
```
Yes, you're right.  It is the same if we just do the following, because uniqueValidator won't have semantic="atomic".
```
if lm.semantic != "" {
...
```
I was thinking that we should keep `&& lm.semantic != semanticAtomic` to protect any future logic leak? But maybe we don't need to do it.

thockin replied:
Up to you - I just wantedf to make sure there was not some corner case that I missed and we could somehow get into that state.

yongruilin replied:
I would just keep it.

yongruilin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/list.go:
```diff
@@ -0,0 +1,426 @@
+	}
+	// TODO: replace with the following once we have a way to either opt-out from this validation
+	// or settle the decision on how to handle the ratcheting cases.
+	// if lm.semantic == semanticMap {
```
We disable unique for listtype=map, but unique=map should be enabled. I need to rephrase the comment.

The reason above line - the reason we disable unique on listmap is that the handwritten in CSR has a too lenient ratcheting logic (allow update from invalid to invalid).

yongruilin replied:
Reworded.

yongruilin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/list.go:
```diff
@@ -0,0 +1,426 @@
+)
+
+// listMeta is shared between list-related validators.
+var listMeta = map[string]*listMetadata{} // keyed by the field or type path
```
Renamed.

yongruilin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/each.go:
```diff
@@ -481,26 +168,25 @@ func (evtv eachValTagValidator) getListValidations(fldPath *field.Path, t *types
+				matchArg = Identifier(validateSemanticDeepEqual)
+			}
 		}
-	default:
```
Oh.. I accidentally dropped it when move the file. I added back this comment.

yongruilin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/validators/list.go:
```diff
@@ -0,0 +1,426 @@
+	RegisterTypeValidator(listValidator{byPath: listMeta})
+
+	// Processing item tags requires the list metadata.
+	RegisterTagValidator(&itemTagValidator{listByPath: listMeta})
```
Updated.

yongruilin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/tags/unique/doc.go:
```diff
@@ -0,0 +1,95 @@
+	// atomic + unique=set combination
+	// +k8s:listType=atomic
+	// +k8s:unique=set
+	AtomicListUniqueSet []Item `json:"atomicListUniqueSet"`
```
Yea.. looks like I lost it when iterated. I removed the duplicate fields and renamed them to make them align.

yongruilin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/tags/unique/doc.go:
```diff
@@ -0,0 +1,95 @@
+	// Test with zero values
+	// +k8s:listType=atomic
+	// +k8s:unique=set
+	SliceWithZeroValues []string `json:"sliceWithZeroValues"`
```
SG, updated.

yongruilin commented on staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/tags/unique/doc_test.go:
```diff
@@ -0,0 +1,240 @@
+		field.Duplicate(field.NewPath("sliceSetField").Index(4), "bbb"),
+		field.Duplicate(field.NewPath("sliceSetField").Index(5), "aaa"),
+		field.Duplicate(field.NewPath("sliceMapField").Index(2), nil),
+		field.Duplicate(field.NewPath("sliceSetFieldWithStruct").Index(2), Item{Key: "key1", Data: "one"}),
```
field path should be sufficient. I changed them to nil.

thockin commented:
I think this looks solid!

The intersections of tests is getting complicated again - we should think about if/how we can make it clearer what EXACTLY each test case is proving.
